### PR TITLE
Improve torch version check

### LINKED ISSUE #29861: Pipeline can't find torch or tensorflow, errors out in notebook.
### System Info

- huggingface_hub version: 0.22.0
- Platform: Windows-11-10.0.22631-SP0
- Python version: 3.12.2
- Running in iPython ?: No
- Running in notebook ?: 
- Running in Google Colab ?: No
- Token path ?: C:\Users\rober\.cache\huggingface\token
- Has saved token ?: False
- Configured git credential helpers: manager
- FastAI: N/A
- Tensorflow: 2.16.1
- Torch: 2.2.1
- Jinja2: 3.1.3
- Graphviz: N/A
- keras: 3.1.1
- Pydot: N/A
- Pillow: N/A
- hf_transfer: N/A
- gradio: N/A
- tensorboard: N/A
- numpy: 1.26.4
- pydantic: N/A
- aiohttp: N/A
- ENDPOINT: https://huggingface.co
- HF_HUB_CACHE: C:\Users\rober\.cache\huggingface\hub
- HF_ASSETS_CACHE: C:\Users\rober\.cache\huggingface\assets
- HF_TOKEN_PATH: C:\Users\rober\.cache\huggingface\token
- HF_HUB_OFFLINE: False
- HF_HUB_DISABLE_TELEMETRY: False
- HF_HUB_DISABLE_PROGRESS_BARS: None
- HF_HUB_DISABLE_SYMLINKS_WARNING: False
- HF_HUB_DISABLE_EXPERIMENTAL_WARNING: False
- HF_HUB_DISABLE_IMPLICIT_TOKEN: False
- HF_HUB_ENABLE_HF_TRANSFER: False
- HF_HUB_ETAG_TIMEOUT: 10
- HF_HUB_DOWNLOAD_TIMEOUT: 10

### Who can help?

_No response_

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

I'm also having a similar problem to https://github.com/huggingface/transformers/issues/27214 on Windows 11, trying to run the code from one of the tutorials in Jupyter Lab:

```
from transformers import pipeline

translator = pipeline(task="translation", model="facebook/nllb-200-distilled-600M")
```
This gives me a similar error to that issue: `RuntimeError: At least one of TensorFlow 2.0 or PyTorch should be installed.`

Following the suggestion in that issue, if I run:
```
import torch

from transformers import is_torch_available

print(torch.__version__)
print(is_torch_available())
```
I get:
```
2.2.1+cpu
False
```
I went into the repo and reproduced the code (i.e. copy-pasted it) in import_utils.py that defines `is_torch_available()` into a notebook and set the environment variables:
```
USE_TORCH=1
USE_TF=0
```
This caused my reproduced code to return `true` for `is_torch_available()`, but the actual call that's being made in my notebook is returning `false` for whatever reason. The notebook running the reproduced code can find the environment variables, and it can find the package, and it can identify the package version (2.2.1), but this doesn't work when the library version of the function is called.

### Expected behavior

It should return `true` for `is_torch_available()`, leading to the correct execution of the line from the tutorial.

### PR DESCRIPTION
**Pull request by Rocketknight1**
`import_utils.py` has a lot of functions for checking when packages are available, and also what their version is. However, some problems can occur because `_is_package_available` uses `importlib.metadata` to check versions, but this checks the version of the [distribution package, not the import package](https://packaging.python.org/en/latest/discussions/distribution-package-vs-import-package/). Therefore, it fails when the distribution package cannot be correctly located, even if `torch` is available.

I reworked the test in the following ways:

1) Stop reporting any package as unavailable just because we can't find its version, unless the version was specifically requested
2) Modify the `torch` check specifically to import torch and check its version directly, in the case when `torch` is importable but the distribution package version cannot be determined.

Fixes #29861

### CODE REVIEW

File: src/transformers/utils/import_utils.py [hunk 1]
```diff
@@ -57,12 +57,13 @@ def _is_package_available(pkg_name: str, return_version: bool = False) -> Union[
...
                         package_exists = True
                     else:
-                        package_exists = False
+                        package_version = "N/A"
                 except ImportError:
                     # If the package can't be imported, it's not available
-                    package_exists = False
-            else:
-                # For packages other than "torch", don't attempt the fallback and set as not available
+                    package_version = "N/A"
```
Comment (ydshieh) [inline src/transformers/utils/import_utils.py:63]: At this line, the torch can't be imported, no? So why we can't keep `package_exists = False`?

**Comment by ydshieh** [comment]
The user reported the issue seems use an installed torch 2.2.1 (I guess this is `distribution package`?). Do you know which line the check fails and cause `_is_package_available` to return `False`?

File: src/transformers/utils/import_utils.py (hunk 1 above)
Comment (Rocketknight1) [reply src/transformers/utils/import_utils.py:63]: `torch` can be imported at this line! We only get here if `importlib.util.find_spec("torch")` is True, right?

**Comment by Rocketknight1** [comment]
@ydshieh I think the check fails when you can `import torch`, which means `importlib.util.find_spec("torch") is not None` but `importlib.metadata.version("torch")` raises an error. In other words, the check fails when the **import package** is present but the **distribution package** is not.

**Comment by ydshieh** [comment]
I have some doubts because: it seems the issue author has a normal torch installation (`2.2.1+cpu`) which is the same as our CircleCI and yet they get the error but our CircleCI is fine.

It's unclear (to me) that `importlib.util.find_spec` and `importlib.metadata.version` would behave differently between their env. and our CircleCI env.

**Comment by Rocketknight1** [comment]
I'll see what I can do! I suspect it's something to do with the specific package manager in use.

@robertdecaire can you give us some information about how exactly Python and Torch are installed on your system? Which distribution of Python are you using, and did you install `torch` with their package manager, or with another package manager like `conda` or `pip`?

**Comment by ydshieh** [comment]
> I'll see what I can do! I suspect it's something to do with the specific package manager in use.
> 
> @robertdecaire can you give us some information about how exactly Python and Torch are installed on your system? Which distribution of Python are you using, and did you install `torch` with their package manager, or with another package manager like `conda` or `pip`?

In particular, you mentioned you are running on  Jupyter Lab. Are you able to reproduce with the same issue with the command line directly?

**Comment by ydshieh** [comment]
@Rocketknight1 Let's wait until the author of the original opened issue provide some more information regarding the env. to reproduce.

**Comment by Rocketknight1** [comment]
Tempted not to merge this - we haven't seen other reports of the problem since I made the PR,  so I'm not sure we want to complicate the import machinery with a heuristic. It may only be a problem with very idiosyncratic environments. I'll close it for now, but we can reopen it if needed!

**Pull request closed without merging**